hub/util/merge.py

```diff
@@ -0,0 +1,321 @@
+    target_commit_id = dataset.pending_commit_id
+    dataset.checkout(original_id)
+    return target_commit_id
+
+
+def get_changes_commit_ids_for_node(
+    dataset, tensor_name: str, commit_node: Optional[CommitNode], lca_node: CommitNode
+):
+    changes_commit_map = defaultdict(list)
+    current_node = commit_node
+    while current_node and current_node.commit_id != lca_node.commit_id:
+        commit_id = current_node.commit_id
+        if current_node.is_merge_node:
+            changes = get_changes_commit_ids_for_node(
+                dataset, tensor_name, current_node.merge_parent, lca_node
+            )
+            for idx in changes:
+                changes_commit_map[idx].extend(changes[idx])
+        else:
+            try:
+                diff_key = get_tensor_commit_diff_key(tensor_name, commit_id)
+                diff: CommitDiff = dataset.storage.get_hub_object(diff_key, CommitDiff)
+                data_updated = sorted(diff.data_updated)
+                id_tensor_name = get_sample_id_tensor_key(tensor_name)
+                id_tensor = dataset[id_tensor_name]
+                for idx in data_updated:
+                    sample_id = id_tensor[idx].numpy()[0]
+                    changes_commit_map[sample_id].append(commit_id)
+            except KeyError:
+                pass
+        current_node = current_node.parent
+    return changes_commit_map
+
+
+def delete_tensors(dataset, tensor_names: Set[str]):
+    for tensor_name in tensor_names:
+        dataset.delete_tensor(tensor_name)
+
+
+def copy_new_tensors(dataset, target_dataset, tensor_names: Set[str]):
+    for tensor_name in tensor_names:
+        target_tensor = target_dataset[tensor_name]
+        htype = target_tensor.meta.htype
+        sample_compression = target_tensor.meta.sample_compression
+        chunk_compression = target_tensor.meta.chunk_compression
+        dataset.create_tensor(
+            tensor_name,
+            htype=htype,
+            sample_compression=sample_compression,
+            chunk_compression=chunk_compression,
+        )
+        new_tensor = dataset[tensor_name]
+        for item in target_tensor:
+            new_tensor.append(item)
+        new_tensor.info.update(target_tensor.info)
+
+
+def merge_common_tensors(
+    tensor_names: Set[str],
+    dataset,
+    target_dataset,
+    original_node: CommitNode,
+    target_node: CommitNode,
+    lca_node: CommitNode,
+    conflict_resolution: Optional[str] = None,
+):
+    check_common_tensor_mismatches(dataset, target_dataset, tensor_names)
+    new_samples_dict: Dict[str, List[int]] = {}
+    conflict_samples_dict: Dict[str, List[Tuple[int, int]]] = {}
+
+    for tensor_name in tensor_names:
+        process_tensor(
+            tensor_name,
+            dataset,
+            target_dataset,
+            original_node,
+            target_node,
+            lca_node,
+            new_samples_dict,
+            conflict_samples_dict,
```

Comment:
in a for loop this dict in place changes, but you would never say reading the either the function or the loop separately, better to return it.